eldritch/assets.py

```diff
@@ -273,6 +273,44 @@ def get_trigger(self, event, owner, state):
     return None
 
 
+class BlessingOrCurse(Card):
+  def __init__(self, kind, idx):
+    assert kind in ["Blessing", "Curse"]
+    super().__init__(kind, idx, "specials", {}, {})
+    self.opposite = "Curse" if kind == "Blessing" else "Blessing"
+    self.must_roll = False
+
+  def get_trigger(self, event, owner, state):
+    if isinstance(event, events.UpkeepActions) and event.character == owner:
+      if self.must_roll:
+        roll = events.DiceRoll(owner, 1)
+        return events.Conditional(
+            owner, roll, "sum",
+            {1: events.DiscardSpecific(owner, self), 2: events.Nothing()}
+        )
+      self.must_roll = True
+
+    if isinstance(event, events.KeepDrawn) and self.name in event.kept:
+      selves = [p for p in event.character.possessions if p.name == self.name]
+      _, *duplicates = sorted(selves, key=lambda x: x.must_roll)
+      if duplicates:
+        return events.DiscardSpecific(event.character, duplicates)
```

Comment:
This may be problematic. There are a few cards that say "if you lose your blessing, ...". Might make more sense to use an interrupt to change the KeepDrawn event? There's also the Nun's personal story, which adds a token each time a player is blessed. Do we think that story should advance if a player is blessed while they already have a blessing?

If no, then you might want to make this an interrupt - interrupt a KeepDrawn and then replace it with a different event (either change tokens, or discard specific). If yes, then we may need to make bless/curse its own event type to deal with this (or hard-code it into KeepDrawn, but I think that will get messy).

Reply:
I would think that re-blessing a player would advance the nun's story, but should not count as them losing their blessing. From the FAQ:
> Treat the card as if it were newly acquired, so do not roll during the next upkeep to see if you lose it.

One way to implement this could be to make an empty subclass of DiscardSpecific LoseBlessingOrCurse, and then have the logic be:

- LoseBlessingOrCurse if you roll a 1
- LoseBlessingOrCurse if you draw the opposite
- DiscardSpecific(event.character, duplicates)

That way, cards that care that you're losing your blessing can look for the LoseBlessingOrCurse, but the behavior is ultimately the same.